background_scripts/marks.coffee

```diff
@@ -1,34 +1,80 @@
+        scrollY: req.scrollY
+      chrome.storage.sync.set item
 
-chrome.tabs.onRemoved.addListener (tabId, removeInfo) ->
-  # XXX(jez): what about restored tabs?
-  removeMarksForTab tabId
+  # Goto a global mark.  We try to find the original tab.  If we can't find that, then we try to find another
+  # tab with the original URL, and use that.  And if we can't find such an existing tab, then we create a new
+  # one.  Whichever of those we do, we then set the scroll position to the original scroll position.
+  goto: (req, sender) ->
+    chrome.storage.local.get "vimiumSecret", (items) =>
+      vimiumSecret = items.vimiumSecret
+      key = @getLocationKey req.markName
+      chrome.storage.sync.get key, (items) =>
+        markInfo = items[key]
+        if not markInfo
+          # The mark is not defined.
+          chrome.tabs.sendMessage sender.tab.id,
+            name: "showHUDforDuration",
+            text: "Global mark not set: '#{req.markName}'."
+            duration: 1000
+        else if markInfo.vimiumSecret != vimiumSecret
+          # This is a different Vimium instantiation, so markInfo.tabId is definitely out of date.
+          @focusOrLaunch markInfo
+        else
+          # Check whether markInfo.tabId still exists.  According to here (https://developer.chrome.com/extensions/tabs),
+          # tab Ids are unqiue within a Chrome session.  So, if we find a match, we can use it.
+          chrome.tabs.get markInfo.tabId, (tab) =>
+            if not chrome.runtime.lastError and tab?.url and markInfo.url == @getBaseUrl tab.url
+              # The original tab still exists.
+              @gotoPositionInTab markInfo
+            else
+              # The original tab no longer exists.
+              @focusOrLaunch markInfo
 
-removeMarksForTab = (id) ->
-  for markName, mark of marks
-    if mark.tabId is id
-      delete marks[markName]
+  # Focus an existing tab and scroll to the given position within it.
+  gotoPositionInTab: ({ tabId, scrollX, scrollY, markName }) ->
+    chrome.tabs.update tabId, { selected: true }, ->
+      chrome.tabs.sendMessage tabId,
+        { name: "setScrollPosition", scrollX: scrollX, scrollY: scrollY }, ->
+          chrome.tabs.sendMessage tabId,
+            name: "showHUDforDuration",
+            text: "Jumped to global mark '#{markName}'."
+            duration: 1000
 
-root.goto = (req, sender) ->
-  mark = marks[req.markName]
-  chrome.tabs.update mark.tabId, selected: true
-  chrome.tabs.sendMessage mark.tabId,
-    name: "setScrollPosition"
-    scrollX: mark.scrollX
-    scrollY: mark.scrollY
-  chrome.tabs.sendMessage mark.tabId,
-    name: "showHUDforDuration",
-    text: "Jumped to global mark '#{req.markName}'"
-    duration: 1000
+  # The tab we're trying to find no longer exists.  We either find another tab with a matching URL and use it,
+  # or we create a new tab.
+  focusOrLaunch: (markInfo) ->
+    chrome.windows.getAll { populate: true }, (windows) =>
```

Comment:
This would be equivalent to

```coffee
chrome.tabs.query {url: [markInfo.url, markInfo.url + "#*"]}, (tabs) ->
  if tabs.length > 0
     # We have a matching tab: use it.
    @gotoPositionInTab extend markInfo, tabId: tabs[0].id
  else
    # There is no existing matching tab, we'll have to create one.
    chrome.tabs.create { url: @getBaseUrl markInfo.url }, (tab) =>
      # Note. tabLoadedHandlers is defined in "main.coffee".  The handler below will be called when the tab
      # is loaded, its DOM is ready and it registers with the background page.
      tabLoadedHandlers[tab.id] = => @gotoPositionInTab extend markInfo, tabId: tab.id
```

which skips us checking each window and tab manually.